mava/systems/ppo/sebulba/ff_ippo.py

```diff
@@ -11,3 +11,734 @@
+            actor = threading.Thread(
+                target=rollout,
+                args=(
+                    jax.device_put(key, devices[d_id]),
+                    config,
+                    pipeline,
+                    params_sources[-1],
+                    apply_fns,
+                    d_id,
+                    seeds,
+                    actors_lifetime,
+                ),
+                name=f"Actor-{thread_id + d_idx * config.arch.n_threads_per_executor}",
+            )
+            actor.start()
+            actor_threads.append(actor)
+
+    learner_queue: Queue = Queue()
+    threading.Thread(
+        target=learner,
+        name="Learner",
+        args=(learn, learner_state, config, learner_queue, pipeline, params_sources),
+    ).start()
+
+    max_episode_return = -jnp.inf
+    best_params = unreplicated_inital_params.actor_params
+
+    # This is the main loop, all it does is evaluation and logging.
+    # Acting and learning is happening in their own threads.
+    # This loop waits for the learner to finish an update before evaluation and logging.
+    for eval_step in range(config.arch.num_evaluation):
+        # Get the next set of params and metrics from the learner
+        episode_metrics, train_metrics, learner_state, times_dict = learner_queue.get()
+
+        t = int(steps_per_rollout * (eval_step + 1))
+        times_dict["timestep"] = t
+        logger.log(times_dict, t, eval_step, LogEvent.MISC)
+
+        episode_metrics, ep_completed = get_final_step_metrics(episode_metrics)
+        episode_metrics["steps_per_second"] = steps_per_rollout / times_dict["single_rollout_time"]
+        if ep_completed:
+            logger.log(episode_metrics, t, eval_step, LogEvent.ACT)
+
+        logger.log(train_metrics, t, eval_step, LogEvent.TRAIN)
+
+        unreplicated_actor_params = flax.jax_utils.unreplicate(learner_state.params.actor_params)
+        key, eval_key = jax.random.split(key, 2)
+        eval_metrics = evaluator(unreplicated_actor_params, eval_key, {})
+        logger.log(eval_metrics, t, eval_step, LogEvent.EVAL)
+
+        episode_return = jnp.mean(eval_metrics["episode_return"])
+
+        if save_checkpoint:
+            # Save checkpoint of learner state
+            checkpointer.save(
+                timestep=steps_per_rollout * (eval_step + 1),
+                unreplicated_learner_state=learner_state,
+                episode_return=episode_return,
+            )
+
+        if config.arch.absolute_metric and max_episode_return <= episode_return:
+            best_params = copy.deepcopy(unreplicated_actor_params)
+            max_episode_return = episode_return
+
+    evaluator_envs.close()
+    eval_performance = float(jnp.mean(eval_metrics[config.env.eval_metric]))
+
+    # Make sure all of the actors are done befor closing the pipeline
+    actors_lifetime.stop()
+    for actor in actor_threads:
+        actor.join()
+    pipeline_lifetime.stop()
+    params_sources_lifetime.stop()
```

Comment:
can we make sure all threads are done
```suggestion
    pipeline_lifetime.stop()
    pipeline.join()
    params_sources_lifetime.stop()
    for params_source in params_sources:
        params_source.join()
```